Public/Src/Cache/ContentStore/Library/Service/Grpc/GrpcContentServer.cs

```diff
@@ -153,97 +156,174 @@ public async Task<GetStatsResponse> GetStatsAsync(GetStatsRequest request, Cance
             };
         }
 
+        private async Task<OpenStreamResult> GetFileStreamAsync (Context context, ContentHash hash)
+        {
+            // For now, take the first store. Change this to interate through all stores
+            // until content is found.
+            IStreamStore store = _contentStoreByCacheName.First().Value as IStreamStore;
+
+            if (store == null)
+            {
+                return new OpenStreamResult("No content store available.");
+            }
+
+            OpenStreamResult result = await store.StreamContentAsync(context, hash);
+
+            return result;
+        }
+
         /// <summary>
         /// Implements a copy file request.
         /// </summary>
         public async Task CopyFileAsync(CopyFileRequest request, IServerStreamWriter<CopyFileResponse> responseStream, ServerCallContext context)
         {
-            DateTime startTime = DateTime.UtcNow;
-
             try
             {
                 LogRequestHandling();
 
-                var cacheContext = new Context(new Guid(request.TraceId), _logger);
-
-                if (!_nameByDrive.TryGetValue(request.Drive, out string name))
-                {
-                    await responseStream.WriteAsync(
-                        new CopyFileResponse
-                        {
-                            Header = new ResponseHeader(startTime, false, (int)CopyFileResult.ResultCode.SourcePathError, $"'{request.Drive}' is an invalid cache."),
-                        });
-                }
+                // Get the content stream.
+                Context cacheContext = new Context(new Guid(request.TraceId), _logger);
+                HashType type = (HashType) request.HashType;
+                ContentHash hash = request.ContentHash.ToContentHash((HashType)request.HashType);
+                OpenStreamResult result = await GetFileStreamAsync(cacheContext, hash);
 
-                if (!_contentStoreByCacheName.TryGetValue(name, out var cache))
+                using (result.Stream)
                 {
-                    await responseStream.WriteAsync(
-                        new CopyFileResponse
-                        {
-                            Header = new ResponseHeader(startTime, false, (int)CopyFileResult.ResultCode.SourcePathError, $"'{name}' is an invalid cache name."),
-                        });
-                }
+                    // Figure out response headers.
+                    CopyCompression compression = CopyCompression.None;
+                    Metadata headers = new Metadata();
+                    headers.Add("ContentHash", hash.ToString());
+                    switch (result.Code)

```

Comment:
maybe it make sense to add a default section that will throw.

Reply:
Fixed.